---
title: "Standard Analysis of bulk RNA-seq"
author: <a href="https://chela-james.github.io/"> <h3> Chela James </h3> </a> \newline Cancer Institute, UCL, UK
date: 15 Aprl 2019
output:
  bookdown::html_document2:
    toc: true
    toc_float: true
    toc_depth: 3
    number_sections: true
    theme: united
    highlight: tango
    df_print: paged
    code_folding: hide
    code_download: true
    self_contained: true
    keep_md: false
    encoding: "UTF-8"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
htmltools::tagList(rmarkdown::html_dependency_font_awesome())
knitr::opts_chunk$set(eval = TRUE, cache = FALSE, warning = FALSE, message = FALSE)
```


```{r,  out.width = "100%", echo = FALSE, warning = FALSE}
library(knitr)
library(ggplot2)
library(sva)
library(dplyr)
library(preprocessCore)
library(ggrepel)
library(Biobase)
library(tibble)
library(scater)
library(gghighlight)
library(annotables)
library(DESeq2)
library(fgsea)
library(DT)
library(purrr)
library(stringr)
library(ggrepel)
library(hypeR)
```



```{r}
knitr::opts_chunk$set(echo = FALSE, cache = TRUE, message = FALSE, warning = FALSE)
```

# Transcriptional profile PCA and differential gene expression (DGA)

## RNA-seq

RNA was isolated and sequenced from the various cell types.

## Sample summary

```{r}
sample_table <- readxl::read_xlsx('Key.xlsx')
sample_table <- data.frame(sample_table)
colnames(sample_table) <- c("index", "sample_name", "origin", "donor", "day", "protocol", "population", "colour")

datatable(sample_table,options=list(columnDefs = list(list(visible=FALSE, targets=8)),   pageLength = 29)) %>% 
  formatStyle( "colour",target = 'row', backgroundColor = styleEqual(sample_table$colour ,sample_table$colour))
  
```

### Using [scater](https://bioconductor.org/packages/release/bioc/html/scater.html) normalization 

```{r}

# read in the counts matrix (in this instance generated by QoRTs)

big_counts <- read.table('ariane_qorts_counts_array.txt',stringsAsFactors = FALSE, header = TRUE)

count_matrix <- as.matrix(big_counts)

sce <- SingleCellExperiment(list(counts = count_matrix),
                            colData = sample_table)
sce <- sce[rowSums(counts(sce)) >  0,]
sce <- calculateQCMetrics(sce)

```

```{r}
## prepare total count and total features data

my_df <- data.frame("sample_id" = colnames(sce), "total_counts" = sce$total_counts,
                    "total_features" = sce$total_features_by_counts )
ggplot(my_df,aes(total_counts)) + geom_histogram(bins = 29) + 
  theme_minimal() + ggtitle("Histogram of total counts") + ylab("number of samples") + xlab("total counts")
```

```{r}
ggplot(my_df,aes(total_features)) + geom_histogram(bins = 29) + 
  theme_minimal() + ggtitle("Histogram of total features") + ylab("number of samples") + xlab("total features")

```

## DEseq2 PCA

```{r}
my_palette <- c("burlywood" = "burlywood", "chocolate" =  "chocolate" , "honeydew" = "honeydew",  "lightblue" = "lightblue",  
 "lightskyblue" = "lightskyblue",  "moccasin" = "moccasin",  "olivedrab" = "olivedrab",  "sandybrown" = "sandybrown",
 "white" = "white")

my_coldata <- sample_table

my_coldata$index <- c(paste0("A0", sample_table$index[1:9]), paste0("A", sample_table$index[10:length(sample_table$index)]))
rownames(my_coldata) <- my_coldata$index
big_counts <- big_counts[, my_coldata$index]
dds <- DESeqDataSetFromMatrix(big_counts, colData = my_coldata, design = ~colour)

my_vst <- vst(dds)

pcaData <- DESeq2::plotPCA(my_vst, intgroup = c("colour","origin","index"),returnData = TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))

ggplot(pcaData,aes(PC1,PC2,label = index)) +
        geom_point(size = 4, aes(shape = origin, fill = colour), colour  = "black") +
        guides(fill = FALSE ) +
        scale_shape_manual(values = c(21,24)) +
        scale_fill_manual(values = my_palette) +
        xlab(paste0("PC1: ",percentVar[1],"% variance")) +
        ylab(paste0("PC2: ",percentVar[2],"% variance")) + geom_text_repel(show.legend = FALSE) + 
        theme_minimal() + ggtitle(label = "PCA using counts data")
```



## DEseq2 PCA (dupRadar)

```{r}

dupRbig_counts <- read.table('ariane_dupradar_counts_matrix.txt',stringsAsFactors = FALSE, header = TRUE)

dupRbig_counts <- big_counts[, my_coldata$index]

dupRdds <- DESeqDataSetFromMatrix(dupRbig_counts, colData = my_coldata, design = ~colour)
dupRmy_vst <- vst(dupRdds)

dupRpcaData <-  DESeq2::plotPCA(my_vst, intgroup = c("colour","origin","index"),returnData = TRUE)
dupRpercentVar <- round(100 * attr(dupRpcaData, "percentVar"))

ggplot(pcaData,aes(PC1,PC2,label = index)) +
        geom_point(size = 4, aes(shape = origin, fill = colour), colour  = "black") +
        guides(fill = FALSE ) +
        scale_shape_manual(values = c(21,24)) +
        scale_fill_manual(values = my_palette) +
        xlab(paste0("PC1: ",percentVar[1],"% variance")) +
        ylab(paste0("PC2: ",percentVar[2],"% variance")) + geom_text_repel(show.legend = FALSE) + 
        theme_minimal() + ggtitle(label = "PCA using counts data")
```


## PCA GSEA

### PC1 GSEA

```{r}
HALLMARK <- msigdb_download_one(species="Homo sapiens", category="H")

pca1_gsea = function(object) {
  # calculate the variance for each gene
  rv <- rowVars(assay(object))

  # select the 1000 top genes by variance
  select <- order(rv, decreasing=TRUE)[seq_len(min(2000, length(rv)))]

  # perform a PCA on the data in assay(x) for the selected genes
  pca <- prcomp(t(assay(object)[select,]))

  PCA1_contrib <- sort(pca$rotation[,1], decreasing = TRUE )
  PCA1_contrib <- data.frame("ensgene" = names(PCA1_contrib), PCA1_contrib)
  PCA1_contrib <- left_join(PCA1_contrib,grch38,by="ensgene")
  PCA1_contrib <- PCA1_contrib[complete.cases(PCA1_contrib$symbol),]
  for_gsea <- PCA1_contrib$PCA1_contrib  
  names(for_gsea) <- PCA1_contrib$symbol
  fgseaRes <- fgsea(HALLMARK, for_gsea, nperm=10000, maxSize = 500)
  fgseaRes <- fgseaRes[fgseaRes$padj < 0.05,]
  plotGseaTable(HALLMARK[fgseaRes$pathway], for_gsea, fgseaRes, gseaParam = 0.4)
  return(PCA1_contrib)
}
pc1_res<-pca1_gsea(my_vst)
```

### Genes Driving PC1

```{r}

datatable(pc1_res,options = list(dom = 'Bfrtip',buttons = c('copy', 'csv', 'excel')))

```


### PC2 GSEA

```{r}

pca2_gsea = function(object) {
  # calculate the variance for each gene
  rv <- rowVars(assay(object))

  # select the 1000 top genes by variance
  select <- order(rv, decreasing=TRUE)[seq_len(min(2000, length(rv)))]

  # perform a PCA on the data in assay(x) for the selected genes
  pca <- prcomp(t(assay(object)[select,]))

  PCA2_contrib <- sort(pca$rotation[,2], decreasing = TRUE )
  PCA2_contrib <- data.frame("ensgene" = names(PCA2_contrib), PCA2_contrib)
  PCA2_contrib <- left_join(PCA2_contrib,grch38,by="ensgene")
  PCA2_contrib <- PCA2_contrib[complete.cases(PCA2_contrib$symbol),]
  for_gsea <- PCA2_contrib$PCA2_contrib  
  names(for_gsea) <- PCA2_contrib$symbol
  fgseaRes <- fgsea(HALLMARK, for_gsea, nperm=10000, maxSize = 500)
  fgseaRes <- fgseaRes[fgseaRes$padj < 0.05,]
  plotGseaTable(HALLMARK[fgseaRes$pathway], for_gsea, fgseaRes, gseaParam = 0.4)
  return(PCA2_contrib)
}
pca2_res <- pca2_gsea(my_vst)
```


### Genes Driving PC2

```{r}
datatable(pca2_res,options = list(dom = 'Bfrtip',buttons = c('copy', 'csv', 'excel')))
```


## combinations

```{r eval = FALSE }
library(gtools)
my_combinations <- data.frame(combinations(n=6,r=2,v=unique(sample_table$GROUP.NAME)),stringsAsFactors = F)

colnames(my_combinations) <- c("first","second")

datatable(my_combinations)
```

## Differential Gene Expression
### 1.

```{r eval = FALSE}
#deseq <- DESeq(dds)
#saveRDS(deseq,"vt40_deseq.rds")
deseq <- readRDS("vt40_deseq.rds")

DGE <- function(c1,c2){ 
res <- results(deseq,contrast = c("GROUP.NAME", c1, c2))
resOrdered <- res[order(res$padj),]
resOrdered <- resOrdered[complete.cases(resOrdered),]
resOrdered <- data.frame(resOrdered)
resOrdered <- left_join(rownames_to_column(resOrdered,var = "ensgene"), grch38, by = "ensgene") %>% select(symbol,log2FoldChange,padj)

res_for_table <- resOrdered  %>% dplyr::filter(padj < 0.1)

return(list(resOrdered=resOrdered,res_for_table=res_for_table))
}

my_res<-DGE(my_combinations$first[1],my_combinations$second[1])
datatable(my_res$res_for_table, extensions = 'Buttons',
          caption = paste(my_combinations$first[1]," vs ",my_combinations$second[1]),
          options = list(dom = 'Bfrtip',
                         buttons = c('copy', 'csv', 'excel')))
```


```{r eval = FALSE}
plot_gsea <- function(res){
res <- res[complete.cases(res),]
res$fcSign <- sign(res$log2FoldChange)
res$logP <- -log10(res$padj)
res$metric <- res$logP/res$fcSign
y<-res[,c("symbol", "metric")]
geneList <- y$metric
names(geneList) <- y$symbol
geneList <- geneList[order(geneList, decreasing = T)]
fgseaRes <- fgsea(hallmark, geneList, nperm=10000, maxSize = 500)
fgseaRes <- fgseaRes[fgseaRes$padj < 0.05,]
plotGseaTable(hallmark[fgseaRes$pathway], geneList, fgseaRes, gseaParam = 0.4)
}

plot_gsea(my_res$resOrdered)
```

### 2.

```{r eval = FALSE}
my_res<-DGE(my_combinations$first[2],my_combinations$second[2])
datatable(my_res$res_for_table, extensions = 'Buttons',
          caption = paste(my_combinations$first[2]," vs ",my_combinations$second[2]),
          options = list(dom = 'Bfrtip',buttons = c('copy', 'csv', 'excel')))
```


```{r eval = FALSE}
plot_gsea(my_res$resOrdered)
```

### 3.

```{r eval = FALSE}
my_res<-DGE(my_combinations$first[3],my_combinations$second[3])
datatable(my_res$res_for_table, extensions = 'Buttons',
          caption = paste(my_combinations$first[3]," vs ",my_combinations$second[3]),
          options = list(dom = 'Bfrtip',buttons = c('copy', 'csv', 'excel')))
```


```{r eval = FALSE}
plot_gsea(my_res$resOrdered)
```

### 4.

```{r eval = FALSE}
my_res<-DGE(my_combinations$first[4],my_combinations$second[4])
datatable(my_res$res_for_table, extensions = 'Buttons',
          caption = paste(my_combinations$first[4]," vs ",my_combinations$second[4]),
          options = list(dom = 'Bfrtip',buttons = c('copy', 'csv', 'excel')))
```


```{r eval = FALSE}
plot_gsea(my_res$resOrdered)
```

### 5.

```{r eval = FALSE}
my_res<-DGE(my_combinations$first[5],my_combinations$second[5])
datatable(my_res$res_for_table, extensions = 'Buttons',
          caption = paste(my_combinations$first[5]," vs ",my_combinations$second[5]),
          options = list(dom = 'Bfrtip',buttons = c('copy', 'csv', 'excel')))
```


```{r eval = FALSE}
plot_gsea(my_res$resOrdered)
```

### 6.

```{r eval = FALSE}
my_res<-DGE(my_combinations$first[6],my_combinations$second[6])
datatable(my_res$res_for_table, extensions = 'Buttons',
          caption = paste(my_combinations$first[6]," vs ",my_combinations$second[6]),
          options = list(dom = 'Bfrtip',buttons = c('copy', 'csv', 'excel')))
```


```{r eval = FALSE}
plot_gsea(my_res$resOrdered)
```

### 7.

```{r eval = FALSE}
my_res<-DGE(my_combinations$first[7],my_combinations$second[7])
datatable(my_res$res_for_table, extensions = 'Buttons',
          caption = paste(my_combinations$first[7]," vs ",my_combinations$second[7]),
          options = list(dom = 'Bfrtip',buttons = c('copy', 'csv', 'excel')))
```


```{r eval = FALSE}
plot_gsea(my_res$resOrdered)
```

### 8.

```{r eval = FALSE}
my_res<-DGE(my_combinations$first[8],my_combinations$second[8])
datatable(my_res$res_for_table, extensions = 'Buttons',
          caption = paste(my_combinations$first[8]," vs ",my_combinations$second[8]),
          options = list(dom = 'Bfrtip',buttons = c('copy', 'csv', 'excel')))
```


```{r eval = FALSE}
plot_gsea(my_res$resOrdered)
```

### 9.

```{r eval = FALSE}
my_res<-DGE(my_combinations$first[9],my_combinations$second[9])
datatable(my_res$res_for_table, extensions = 'Buttons',
          caption = paste(my_combinations$first[9]," vs ",my_combinations$second[9]),
          options = list(dom = 'Bfrtip',buttons = c('copy', 'csv', 'excel')))
```


```{r eval = FALSE}
plot_gsea(my_res$resOrdered)
```


### 10.

```{r eval = FALSE}
my_res<-DGE(my_combinations$first[10],my_combinations$second[10])
datatable(my_res$res_for_table, extensions = 'Buttons',
          caption = paste(my_combinations$first[10]," vs ",my_combinations$second[10]),
          options = list(dom = 'Bfrtip',buttons = c('copy', 'csv', 'excel')))
```


```{r eval = FALSE}
plot_gsea(my_res$resOrdered)
```


### 11.

```{r eval = FALSE}
my_res<-DGE(my_combinations$first[11],my_combinations$second[11])
datatable(my_res$res_for_table, extensions = 'Buttons',
          caption = paste(my_combinations$first[11]," vs ",my_combinations$second[11]),
          options = list(dom = 'Bfrtip',buttons = c('copy', 'csv', 'excel')))
```


```{r eval = FALSE}
plot_gsea(my_res$resOrdered)
```



### 12.

```{r eval = FALSE}
my_res<-DGE(my_combinations$first[12],my_combinations$second[12])
datatable(my_res$res_for_table, extensions = 'Buttons',
          caption = paste(my_combinations$first[12]," vs ",my_combinations$second[12]),
          options = list(dom = 'Bfrtip',buttons = c('copy', 'csv', 'excel')))
```


```{r eval = FALSE}
plot_gsea(my_res$resOrdered)
```



### 13.

```{r eval = FALSE}
my_res<-DGE(my_combinations$first[13],my_combinations$second[13])
datatable(my_res$res_for_table, extensions = 'Buttons',
          caption = paste(my_combinations$first[13]," vs ",my_combinations$second[13]),
          options = list(dom = 'Bfrtip',buttons = c('copy', 'csv', 'excel')))
```


```{r eval = FALSE}
plot_gsea(my_res$resOrdered)
```


### 14.


```{r eval = FALSE}
my_res<-DGE(my_combinations$first[14],my_combinations$second[14])
datatable(my_res$res_for_table, extensions = 'Buttons',
          caption = paste(my_combinations$first[14]," vs ",my_combinations$second[14]),
          options = list(dom = 'Bfrtip',buttons = c('copy', 'csv', 'excel')))
```


```{r eval = FALSE}
plot_gsea(my_res$resOrdered)
```


### 15.


```{r eval = FALSE}
my_res<-DGE(my_combinations$first[15],my_combinations$second[15])
datatable(my_res$res_for_table, extensions = 'Buttons',
          caption = paste(my_combinations$first[15]," vs ",my_combinations$second[15]),
          options = list(dom = 'Bfrtip',buttons = c('copy', 'csv', 'excel')))
```


```{r eval = FALSE}
plot_gsea(my_res$resOrdered)
```
